### PR TITLE
Differentiate retry and halt error and retry failed compaction only on retriable error

### PR DESCRIPTION
**Pull request by alexqyle**
**What this PR does**:

Currently, Cortex would retry any error from compaction. However, some error would be for sure fail again on retry. This PR adds logic to differentiate halt and retriable error (from Thanos). Cortex compactor should only retry compaction on retriable errors. Also, added metrics to count halt and retriable errors separately 

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/compactor.go [hunk 1]
```diff
@@ -792,10 +792,20 @@ func (c *Compactor) compactUserWithRetries(ctx context.Context, userID string) e
 		if lastErr == nil {
 			return nil
 		}
+		if ctx.Err() != nil {
+			level.Warn(util_log.WithContext(ctx, c.logger)).Log("msg", "received context error during compaction", "org_id", "err", ctx.Err())
+			return ctx.Err()
+		}
 		if c.isCausedByPermissionDenied(lastErr) {
-			level.Warn(c.logger).Log("msg", "skipping compactUser due to PermissionDenied", "user", userID, "err", lastErr)
+			level.Warn(c.logger).Log("msg", "skipping compactUser due to PermissionDenied", "org_id", userID, "err", lastErr)
```
Comment (yeya24) [inline pkg/compactor/compactor.go:800]: Is there a specific reason to change this to `org_id`? I think we use `user` here everywhere in compactor

File: pkg/compactor/compactor.go [hunk 2]
```diff
@@ -792,10 +792,20 @@ func (c *Compactor) compactUserWithRetries(ctx context.Context, userID string) e
 		if lastErr == nil {
 			return nil
 		}
+		if ctx.Err() != nil {
+			level.Warn(util_log.WithContext(ctx, c.logger)).Log("msg", "received context error during compaction", "org_id", "err", ctx.Err())
```
Comment (yeya24) [inline pkg/compactor/compactor.go:796]: Is this log even necessary? context error should only happen when compactor instance is shutting down. For me it is fine to just return the error

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (alexqyle) [reply pkg/compactor/compactor.go:800]: We use `org_id` everywhere: https://github.com/cortexproject/cortex/blob/master/pkg/util/log/wrappers.go#L18

File: pkg/compactor/compactor_metrics.go [hunk 3]
```diff
@@ -159,6 +161,14 @@ func newCompactorMetricsWithLabels(reg prometheus.Registerer, commonLabels []str
 		Name: "cortex_compactor_remaining_planned_compactions",
 		Help: "Total number of plans that remain to be compacted. Only available with shuffle-sharding strategy",
 	}, commonLabels)
+	m.compactionRetryErrors = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
```
Comment (danielblando) [inline pkg/compactor/compactor_metrics.go:164]: We dont have a metric for the PermissionDenied error. Shouldn't we create only one metric and put a reason as label? Then we can expand for the PermissionDenied or other issues in the future.
wdyt?

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (yeya24) [reply pkg/compactor/compactor.go:799]: I believe we have quite a lot of logs using user as key in compactor. Do we plan to change all of them

File: pkg/compactor/compactor_metrics.go (hunk 3 above)
Comment (alexqyle) [reply pkg/compactor/compactor_metrics.go]: Good suggestion. I have merged those two metrics into one and add `type` label for the metric to indicate if it is retriable, halt, or unauthorized error

**Review by danielblando — APPROVED** [review]

File: pkg/compactor/compactor.go (hunk 1 above)
Comment (alexqyle) [reply pkg/compactor/compactor.go]: I reverted the log change for now. Would address it in separate PR. Would utilize the log wrapper where possible in compactor

File: pkg/compactor/compactor_metrics.go [hunk 4]
```diff
@@ -37,12 +37,18 @@ type compactorMetrics struct {
...
-	TimeRangeLabelName = "time_range_milliseconds"
-	ReasonLabelName    = "reason"
+	UserLabelName                 = "user"
+	TimeRangeLabelName            = "time_range_milliseconds"
+	ReasonLabelName               = "reason"
+	CompactionErrorTypesLabelName = "type"
+
+	RetriableError    = "retriable"
+	HaltError         = "halt"
+	UnauthorizedError = "unauthorized"
```
Comment (yeya24) [inline pkg/compactor/compactor_metrics.go:51]: Nit. Shall we keep those consts as private?

**Review by yeya24 — APPROVED** [review]

**Pull request merged**